### DISCUSSION TITLE
Serializers & Permissions Per Action

### DISCUSSION THREAD

**Discussion by ruslan-korneev** [encode/django-rest-framework#8989] [Ideas & Suggestions]
I would like to add these useful mixins, needed when you wanna use a `ViewSet` and different serializers/permissions for different actions, usage should be something like:

```python
class GuitarViewSet(SerializerPerAction, ModelViewSet):
    ...
    action_serializers = {
        "default": GuitarDefaultSerializer,  # this serializer would be default, in this example for actions: create, retrieve, destroy, update, partial-update
        "list": GuitarListSerializer,
        "custom": GuitarCustomSerializer,
    }
    ...

    @action(details=False, ...):
    def custom(self, request, ...):
        ...
```

the same goes for `PermissionPerAction`

**Comment by ruslan-korneev** [comment]
here is a [gist](https://gist.github.com/ruslan-korneev/59077cdbc5b3ec296e11843a27c102f4) i've created with usage and implementation

**Reply by r3dm1ke** [reply]
What are the advantages of using this instead of multiple views?

**Reply by ruslan-korneev** [reply]
We will have less code. A lot of developers, teams, companies use only ViewSets, or only GenericAPIViews, or both, but ViewSet as a very valuable tool does not have a solution to simply add different serializers and permissions. We always have to override `get_serializer_class` or `get_permissions` in that case
